### ISSUE TITLE
add "latest" aggregation

### ISSUE THREAD

**Issue by mejymejy** [SEL-Columbia/bamboo#269]
`latest(<time_index_column>, <value_column>)` evaluates to the latest value in column `value_column` as interpreted by the time series column `time_index_column`

**Comment by pld** [comment]
is `value_column` optional?  defaulting to the value of `time_index_column`.

**Comment by mejymejy** [comment]
i think that makes sense, though for some reason i still want to write it: `latest(<value_column>, <time_index_column>)` in which case we could just always require two arguments...

`latest(<time_index_column>)` essentially gives you a "last updated" field with regards to the `<time_index_column>` so it is something that may be used relatively often.

tl,dr; yeah, let's go with an optional `<value_column>` like you suggest

**Comment by prabhasp** [comment]
my thoughts: <value_column> seems like over-doing it on this one. is there
a concrete use case that motivates that?
while you are at latest, shall we also do earliest, essentially min and max
for a time_index_column?

On Thu, Nov 8, 2012 at 5:28 PM, Mark Johnston notifications@github.comwrote:

> i think that makes sense, though for some reason i still want to write it: latest(<value_column>,
> <time_index_column>) in which case we could just always require two
> arguments...
> 
> latest(<time_index_column>) essentially gives you a "last updated" field
> with regards to the <time_index_column> so it is something that may be
> used relatively often.
> 
> tl,dr; yeah, let's go with an optional <value_column> like you suggest
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/modilabs/bamboo/issues/269#issuecomment-10207836.

## 

Prabhas Pokharel
http://twitter.com/prabhasp
US mobile: +1 347 948 7654
skype/facebook/whatever: prabhasp

**Comment by mejymejy** [comment]
the motivation behind this feature is that we want to create a dataset that represents the current state of entities from observed data about those entities over time.

you are right, a use case may be helpful here:

say we have a survey about water points that is meant to update the status of some basic properties of the water point such as whether or not it has power.  the form will contain at the very least some field that identifies what water point we are talking about `id` , a field telling us the state of the power for that water point `has_power`, and a timestamp for the observation `timestamp`.  our dataset will have lots of observations about water points, including many observations about water point `A`, a few about water point `B` and so on.  in order to know what the current state of these water is we need to somehow distill our dataset into a view that combines the data in a way such that we have the latest value for `has_power` and any other information that we may want to know.

now, a simple way to do this in bamboo would be to group on `id` and somehow aggregate to the "latest" value of any piece of information that you want info.  this would give you a new dataset with one row per `id` (i.e. water point) and columns that contain up-to-date values for fields such as `has_power`.  since we can already group all of the data about the water points by using a groupby on `id`, we just need an aggregation that will look at a given field (`<value_column> = has_power`) and pull the latest value for that field based on some time index (`<time_index_column> = timestamp`).

what is sort of nice here is that since the group is `id` you will have all aggregate information about water points in the same dataset since bamboo will automatically put data for aggregations with the same group in the same frame.  i think that something along these lines is what we want, though there may be similar approaches that make sense...

@prabhasp, what do you think?

**Comment by prabhasp** [comment]
Okay, I get it, thanks for the explanation.

This operation could be broken down to two things:
: `max_index (field)` : returns the index of a maximal element in a column
: `_get_by_index (index, field)` : given a (row-) index, and a field (ie, column-index), return a value.

So `latest(<value_column>, <time_index_column>)` is essentially `_get_by_index(max(<time_index_column>), <value_column>)`. Seems to me that there is value in exposing the internal functions too... agree? 

[Aside: Of course, we'll need to figure out a way to expose the idea of a row_index... do we have a way to do this yet? is it just a record `_id`? Not sure how pandas does things, but R indexes rows by an incremental index, but this can be over-ridden at any time if a `row.names` vector is passed in.]

**Comment by pld** [comment]
When in MongoDB the row index would an `_id`, we have been treating DataSets as... well sets, when they go into pandas they have a default index, which is based on the order of the rows passed to the DataFrame constructor.  It will be consistent given the order of the data pulled from MongoDB is consistent.  So yeah, there is no true index.

We could enforce an internal order by ordering on the MongoDB `_id` before passing the rows to DataFrame.

Let's only build what is going to be used immediately.

**Comment by pld** [comment]
I changed the name of this to "newest", since if we want to add an "oldest" this seems clearer. I named what would be `max_index` `argmax`, since that is what panda's uses internally.

I experimented with a `fetch` aggregation which would be like @prabhasp's

``` python
_get_by_index(max(<time_index_column>), <value_column>)
```

But this would need the parser to support nested aggregations, which turns out to be non-trivial.  I added this issue which should make nesting easier and I think is the direction we should push aggregations in, https://github.com/modilabs/bamboo/issues/325.

In lieu of messing with the parser I have the `NewestAggregation` call the `ArgMaxAggregation` and use its results, something of a compromise.

This could use some testing with real data.
